docs/design/phases.md

```diff
@@ -0,0 +1,110 @@
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - be in the `system:masters` organization
+ - a KubeConfig file for kubelet to use: `/etc/kubernetes/kubelet.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - be in the `system:nodes` organization
+
+The `ca.crt` is also embedded in the KubeConfig files.
+
+### Generate a token.csv file for bootstrapping purposes
+
+There should be:
+ - a `/etc/kubernetes/tokens.csv` file that contains one kubeadm-generated token
+   - the content should be in the form `${token},${user},${uuid},${group}` where:
+     - `${token}` is the full token `123456.0123456789abcdef`
+     - `${user}` is currently `kubeadm-node-csr`, should be changed
+     - `${uuid}` is a random UUID created with `"k8s.io/apimachinery/pkg/util/uuid".NewUUID()`
+     - `${group}` is currently `kubeadm:kubelet-bootstrap`, might be changed
+
+### Bootstrap the control plane
+
+First, see so there is an etcd cluster somewhere:
+ - use an external one if external etcd params are given
+ - spin up one locally if external etcd should not be enabled. It should:
+   - be spun up in a Pod
+   - listen on `localhost:2379`
+   - set `PodSpec.SecurityContext.SELinuxOptions.Type=spc_t`
+   - be at a minimum version `3.0.14`
+   - make a volume out to the host on a given host path
+
+Then, spin up the control plane as static pods:
+ - They are located in `/etc/kubernetes/manifests`
+ - Right now `.json` files, but will probably be `.yaml` files in v1.6
+ - `HostNetwork: true` is present on all Pods since there is no network configured yet
+
+Then kubeadm waits for the control plane to come up as static pods
+
+The master node is patched; the master label is added and the master is tainted.
+
+If self-hosting should be enabled, do a little bit more:
+ - TODO
+
+### Configure the discovery mechanism
+
+There are two options here:
+ - output a kubeconfig file that can be used for bootstrapping
+ - configuring the BootstrapSigner to create the public `cluster-info` ConfigMap in `kube-public`
+
+### Configure the API Objects
+
+ServiceAccounts for `kube-dns` and `kube-proxy` are created in the `kube-system` namespace.
+
+If RBAC is enabled, then ClusterRoleBindings are created for:
+ - the TLS Bootstrap role (`system:node-bootstrapper`), so tokens in the group `kubeadm:kubelet-bootstrap` only can access what's required for TLS Bootstrapping
+ - kube-proxy; the `kube-proxy` ServiceAccount can gets the privileges in the `system:node-proxier` ClusterRole
+ - kube-dns; the `kube-dns` ServiceAccount can gets the privileges in the `system:kube-dns` ClusterRole
+
+In the future; the base layer of configuration for all kubelets in the cluster will be set in this phase.
```

Comment:
Suggested edit:

In the future the base layer of configuration for all kubelets in the cluster will be set in this phase.